### PR TITLE
ddl, util/codec: fix add index OOM and prevent panic in logging

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/27687

Reproducible step:
```sql
drop table if exists t;
create table t (a bigint primary key);
insert into t values (-9078130948872077313);
alter table t add index idx(a);
```

Problem Summary:

This problem may occur in the following conditions:
- The table contains an clustered integer primary key.
  - If it is unsigned, at least one large primary key(> MaxInt64) is explicitly inserted.
  - If it is signed, at least one small negative number is explicitly inserted.
- An adding index job is running.

There is a bug in the current implementation of `NewCommonHandle()`:

https://github.com/pingcap/tidb/blob/37bfd57dac3fda6622bca4b1f0066b29886ac635/kv/key.go#L237-L262

If the argument bytes have a prefix like `0x8203`, `CutOne()` cuts nothing so that the `remain` is unchanged in the loop. This dead loop may cause TiDB OOM quickly because it endlessly `append()` `colEndOffsets`.

### What is changed and how it works?

1. The function `peek()` in the decoding process does not handle the invalid key. It is possible to run into a dead loop and makes TiDB OOM.
2. The function `tryDecodeHandleToString()` is only used for logging, it should be wrapped with `recover()` so that the critical operation is not interrupted.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix an issue that adding index panics by chance.
```

### CODE REVIEW

File: ddl/backfilling.go [hunk 1]
```diff
@@ -425,6 +425,12 @@ func (w *worker) handleReorgTasks(reorgInfo *reorgInfo, totalAddedCount *int64,
 }
 
 func tryDecodeToHandleString(key kv.Key) string {
+	defer func() {
+		r := recover()
```
Comment (hawkingrei) [inline ddl/backfilling.go:429]: It should look like this.
```
if r := recover(); r!= nil {
    // log(r)
}

```

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/backfilling.go [hunk 2]
```diff
@@ -425,6 +425,13 @@ func (w *worker) handleReorgTasks(reorgInfo *reorgInfo, totalAddedCount *int64,
 }
 
 func tryDecodeToHandleString(key kv.Key) string {
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (zimulala) [inline ddl/backfilling.go:429]: Could we use `tidbutil.Recover` instead of this code?

Comment (tangenta) [reply ddl/backfilling.go:429]: But there is no way to print the value of `key`.

If we use `tidbutil.Recover()` and pass a function as the argument, there is no difference from the current implementation.

Comment (zimulala) [reply ddl/backfilling.go:429]: Got it.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**